tinygrad/codegen/kernel.py

```diff
@@ -607,97 +607,82 @@ def name(self) -> str:
-              # for TC=2, we can't do the shapetracker fixup
-              srcs = [fixup_ast(rsrc.src[0]), fixup_ast(rsrc.src[1])]
-            # MUL/SUM instead of WMMA
-            ret = UOp(Ops.REDUCE_AXIS, tc.dtype_out, (srcs[0].alu(BinaryOps.MUL, srcs[1]).cast(tc.dtype_out),), (alu_op, wmma_arg[-1]))
-          else:
-            # real WMMA, use CONTRACT/EXPAND to get the vectorization right
-            wmma_upcast_axes = wmma_arg[-2]
-            wmma_sz = [prod(x[1] for x in l) for l in wmma_upcast_axes]
+        reduce_idx = len(self.bufs) + self.reduceops.index(op) * 2
+
+        def reduced_axes(start, stop):
+          return tuple(i for i in range(start, stop) if resolve(self.sts[reduce_idx].shape[i] != self.sts[reduce_idx + 1].shape[i]))
+        axes = reduced_axes(self.first_reduce + self.group_for_reduces, self.shape_len)
+        grouped_axes = reduced_axes(self.first_reduce, self.first_reduce + self.group_for_reduces)
+
+        if (tc := self.tensor_core) and (self.use_tensor_cores == 1 or self.use_tensor_cores == 3):
+          def fix_st(st, wd_pattern, tcd_pattern):
+            wd, warp_dims = self.global_dims,  tuple(sz for _, sz in tc.threads)
+            tcd, tcd_dims = self.first_upcast, tuple(sz for _, sz in tc.reduce_axes + tc.early_upcast_axes)
+
+            assert st.shape[wd:wd+len(warp_dims)] == warp_dims, f"warp dims wrong: {st.shape[wd:wd+len(warp_dims)]=} != {warp_dims=}"
+            assert st.shape[tcd:tcd+len(tcd_dims)] == tcd_dims, f"tcd dims wrong: {st.shape[tcd:tcd+len(tcd_dims)]=} != {tcd_dims=}"
+            assert tc.expanded_shape is not None
+
+            new_shape = st.shape[:tcd] + tc.expanded_shape + st.shape[tcd+len(tcd_dims):]  # expand the tcd
+            permaxis = list(range(wd)) + [y + (wd if x == 0 else tcd) for x,y in wd_pattern]  + list(range(wd+len(warp_dims),tcd)) + \
+                                         [y + (wd if x == 0 else tcd) for x,y in tcd_pattern] + list(range(tcd+len(tc.expanded_shape),len(new_shape)))
+            return st.reshape(new_shape).simplify().permute(tuple(permaxis)).reshape(st.shape).simplify()
+
+          srcs = list((ret.src[0] if ret.src[0].op is not Ops.CAST else ret.src[0].src[0]).src)
+          for i, tc_pattern in enumerate([tc.st1_pattern, tc.st2_pattern]):
+            if tc_pattern: srcs[i] = srcs[i].view(fix_st(srcs[i].st, *tc_pattern))
+
+            if self.use_tensor_cores == 3:  # for TC=3, emulate the warp addressing with locals
+              local_shape = tuple(1 if i >= self.first_reduce and i < self.first_upcast else s for i, s in enumerate(self.full_shape))
+              st_uop = ShapeTracker.from_shape(local_shape).to_uop()
+              local_buffer = UOp(Ops.DEFINE_LOCAL, tc.dtype_in.ptr(local=True), (), (f"temp{i + 1}", st_uop.arg.real_size()))
+              local_store = UOp.store(local_buffer, st_uop, srcs[i])
+              if tc_pattern: local_store = local_store.view(fix_st(local_store.st, *tc_pattern))
```

Comment:
ohh you're viewing a STORE, you should probably split the STORE upat from LOAD.

Reply:
feel free to modify view_left. but you understand why VIEW of STORE in schedule is invalid right? STORE is the kernel boundary so it's always base.

Reply:
in this case it does you're actually swizzling the store's st

Reply:
Now I'm confused haha

Reply:
I guess it's not much of a swizzle for the stores but a shapetracker replacement. The swizzle is only for the pattern fix for the tensor cores. I think I ended up messing the terms and believing the same thing should apply for both cases when in reality it should (tc pattern fixing and buffers shapetracker replacement).